lib/ddtrace/opentracer/span.rb

```diff
@@ -1,6 +1,83 @@
 module Datadog
   module OpenTracer
+    # OpenTracing adapter for Datadog::Span
     class Span < ::OpenTracing::Span
+      attr_reader \
+        :datadog_span
+
+      def initialize(datadog_span:, span_context:)
+        @datadog_span = datadog_span
+        @span_context = span_context
+      end
+
+      # Set the name of the operation
+      #
+      # @param [String] name
+      def operation_name=(name)
+        datadog_span.name = name
+      end
+
+      # Span Context
+      #
+      # @return [SpanContext]
+      def context
+        @span_context
+      end
+
+      # Set a tag value on this span
+      # @param key [String] the key of the tag
+      # @param value [String, Numeric, Boolean] the value of the tag. If it's not
+      # a String, Numeric, or Boolean it will be encoded with to_s
+      def set_tag(key, value)
+        tap { datadog_span.set_tag(key, value) }
+      end
+
+      # Set a baggage item on the span
+      # @param key [String] the key of the baggage item
+      # @param value [String] the value of the baggage item
+      def set_baggage_item(key, value)
+        tap { context.baggage[key] = value }
```

Comment:
I agree, I don't think the span should be responsible for knowing details about baggage on the span context. That's why I'm a bit confused about this: if the span context manages the baggage, why is there a baggage function on the span at all?

We could implement baggage functions on the span context object that mirror this, and have this span delegate to those. I guess I'm hesitant to do that only because it seems like a modification of the OT spec. But since its all internal anyways, maybe it's okay.

@Kyle-Verhoog What are your thoughts: move this to SpanContext and delegate from Span, or keep as is?

Reply:
I think the delegate makes sense. Also, I forgot to mention this earlier, but we gotta keep in mind the span context instance should be immutable[0]. So setting a key in the baggage for a span context should produce a new span context instance with the key which the span will use.

[0] https://github.com/opentracing/specification/blob/master/specification.md#spancontext

Reply:
I think what this could look like is spancontext.set_baggage_item returns a new span context instance with the new item added and span.set_baggage_item will update its span context reference with this new instance.

Reply:
@Kyle-Verhoog Okay, I think you're right about the immutable part. I think the implementation might need to be a little different: how about we don't implement `SpanContext#set_baggage_item` and force the use of `SpanContext#new` constructor which accepts baggage? Then do as you said, and have the `Span#set_baggage_item` build a new `SpanContext` from an old one?

The more complicated this `SpanContext` management gets, the more it screams "factory pattern" to me. Especially if the tracer has to build these too.

Reply:
Sounds good to me! The `SpanContext#new` should duplicate the baggage that it receives. 👍

Reply:
Alright, I'll go ahead and make this change then.